05-mvc-with-rails/rails-sessions-and-auth-with-devise/readme.md

```diff
@@ -0,0 +1,365 @@
+
+Great that's nice, our view was generated, but more importantly we had a key-value pair get set in this request because of the `testing_session` action in our artists controller:
+
+```ruby
+def testing_session
+  session[:test] = "this was set on the testing_session action"
+end
+```
+
+How do we verify this session key-value pair was set? Let's navigate back to artists index path.
+
+Now we can see `The value of @test is: this was set on the testing_session action`
+
+If we open a new incognito window and navigate to `http://localhost:3000/artists`, we'll see the session isn't set in that browser because it's a brand new session.
+
+When we clear our cookies/session in our browser we can see in our index route that `@test` becomes nil again.
+
+> Now that we have the ability to transfer information from request to request. We can preserve state in our application.
+
+## Setting objects in sessions
+Instead of hardcoding strings into our sessions hash, lets see if we can store objects themselves in the session.
+
+Let's update our show action to set a session. In `app/controllers/artists_controller.rb`:
+
+```ruby
+def index
+  @last_viewed_artist = session[:last_viewed_artist]
+  @artists = Artist.all
+end
+
+def show
+  @artist = Artist.find(params[:id])
+  session[:last_viewed_artist] = @artist
+end
+```
+
+In your `app/views/artists/index.html.erb place the following:
+
+```html
+<h3>Last viewed artist: <%= @artist["name"] %></h3>
+```
+
+Now every time We click on an artists show page it will update the session to contain the most recently viewed artist. Then any time we make a request to the artists index path we can see the most recently viewed artist's name.
+
+> So now we can store objects or attributes of objects in our sessions as well. This may include id's of a user model, which is basically what devise is going to do for us when a user "signs in"
+
+## Deleting parts of the session
+Lets add a new action/route/view for deleting session.
+
+In `app/controllers/artists_controller.rb`:
+
+```ruby
+def deleting_session
+  session.delete(:last_viewed_artist)
+end
+```
+
+In `config/routes.rb`:
+
+```ruby
+get '/artists/deleting_session', to: 'artists#deleting_session'
+```
+
+Lets create `app/views/artists/deleting_session.html.erb` and add the following content:
+
+```html
+<h1>deleting the last viewed artist session key value pair</h1>
+```
+
+When we navigate from the `deleting_session` route back to the artist we can see an error of undefined method.
+
+> This is the sort of thing thats happening when a user "logs out"
+
+Checkout the [testing_sessions branch](https://github.com/ga-dc/tunr_rails_sessions_devise/tree/testing_sessions) for all the code thus far.
+
+### You do
+Take the next 5 minutes to:
+- create a new action in the posts controller that instantiates a session key-value pair
+- save it to an instance variable in the index, and have it display in the index view
+
```

Comment:
noticing a lack of industry examples here - maybe you can use the concept of a shopping cart. sessions are great for these temporary data stores